### ISSUE TITLE
Investigate the performance and consider the removal of usage of .hide() and .show() from Javascript

### ISSUE THREAD

**Issue by chriscct7** [easydigitaldownloads/Easy-Digital-Downloads#3065]
.show() and .hide() are known jQuery performance bottlenecks per https://github.com/jquery/jquery.com/issues/88#issuecomment-72400007. Until they fix the issues with the performance in https://github.com/jquery/jquery/issues/2057 we should consider implementing alternatives if it makes a noticable difference in performance, that way we aren't waiting for jQuery to fix it, which may or may not take a while

**Comment by cklosowski** [comment]
We can investigate, but I think we'll see negligible differences. I'd vote though that we don't change anything from WP-Admin unless WP Core does.

Our front end we can change, but we're going to have to be EXTRA EXTRA careful about ANY method we use in ALL browsers...going back pretty far to be honest.

**Comment by chriscct7** [comment]
I would think addClass is even more compatible across all browsers. We could just define an edd-hidden class and add remove that.

**Comment by cklosowski** [comment]
At that point  you are modifying CSS Classes though and not using an 'on-element' which trumps ANY css class properties, therefore, any extra CSS applied can be overridden by the user. ```.show()``` and ```.hide()``` are the only way to ensure that no CSS styles are affecting the execution as it's on element.

**Comment by chriscct7** [comment]
Let's examine the performance of each, and if we find a bottleneck, then we can start working a solution. No point looking for a solution for a problem we may or may not have :)

**Comment by cklosowski** [comment]
Define ```bottleneck```

**Comment by chriscct7** [comment]
See https://github.com/jquery/jquery.com/issues/88#issuecomment-72400007. In that case it's causing a 700ms delay. Perhaps a better issue, or maybe a new one might be examine all Javascript and find out what our worst performing parts are and see if we can improve them

**Comment by chriscct7** [comment]
https://github.com/ugma/ugma/issues/21 also makes a great point about .show()/.hide()/.toggle() which is that it could have battery ramifications for mobile applications

**Comment by pippinsplugins** [comment]
We don't use these much at all. While I'm certainly not opposed to making something faster, I'm unsure of whether this is a place we should be changing. Might be best to just let jQuery fix it.

**Comment by pippinsplugins** [comment]
@chriscct7 Minor other note on this. Please do not tag issues with so many labels. "Templates" and "Theme Compatibility" do not apply in nay way to use. Use the ones that make sense and only those. Thanks.

**Comment by chriscct7** [comment]
@pippinsplugins what? I only put 1 tag on this, enhancements. Blame @cklosowski for the rest.

**Comment by cklosowski** [comment]
If we are affecting any front end JS In the end both those could affect themes and our own standard templates.

**Comment by chriscct7** [comment]
It should be noted jQuery is actively telling people not to use it. They're removing the use of those functions from their homepage

**Comment by pippinsplugins** [comment]
Proposed alternatives?

**Comment by chriscct7** [comment]
The current jQuery recommendation is to use classes, a recommendation made on a related ticket about how show and hide are breaking some responsive websites: https://github.com/jquery/jquery/issues/1767

**Comment by pippinsplugins** [comment]
There are some places we could do this, but only where classes already exist and cannot have been modified by theme files.

**Comment by Kflash** [comment]
@chriscct7 Hi. I saw you used a referance to my repo - ugma. Anyway. Have you looked at angularJS?You could use the standards-based aria-hidden attribute. Mostly this problem occur if you check for style.visible === true / false. So I mostly solved my issue with checking if the aria attribute are hidden or not. Shouldn't affect anything. There are pro and cons ofc.

I also see that jQuery has problem with responsive stylesheets. I don't have that problem at all in the way I solved my problem. But when I said that, my hide/show/toggle method are just more then jQuery equalent. I'm also dealing with CSS animation and transitions.

Anyway. Have a look at angularJS: https://docs.angularjs.org/api/ngAria

**Comment by chriscct7** [comment]
@kflash We're not interesting in bundling AngularJS into EDD core.

**Comment by chriscct7** [comment]
Just to re-iterate, the point of this issue isn't to find a replacement, its to determine if we suffer from a performance issue because of the reported issue that jQuery has logged. We do not need to look for solutions until after we determine whether or not we have a problem, its a completely moot activity. Once someone has verified that there is in fact a problem, then a new ticket should be opened to explore solutions as it will likely have quite a bit of debate and thought.

**Comment by Kflash** [comment]
@chriscct7 Just a suggestion on alternative solution was using aria attribute. But I see the main issue here is to find out if there is any performance issue. Should be easy enough. But the source code only contain 10% javascript. Where in the code is this possible bottleneck located? I can have a look.

**Comment by chriscct7** [comment]
According to jQuery any use of .show or .hide(). It shouldn't take long to test, I was planning on profiling it this weekend to check. We've got probably 8 to 15 uses of both of them combined.

**Comment by Kflash** [comment]
@chriscct7 I just looked into the javascript code on this webpage, and I think show() / hide() are not the biggest problems comparing to all the re-flows, mis-use of $, slow javascript performance and a ton of extra bytes you could have shaved off with simplifying.

https://developers.google.com/speed/articles/reflow

**Comment by chriscct7** [comment]
Curious what misuse of $ you found

**Comment by Kflash** [comment]
@chriscct7  Example:  https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/master/assets/js/admin-scripts.js#L85-L97

Here is a lot of this $.  This could be shortened and reduced to 1 if you use the event target.

Example $(this) could be the event.target.  Then you can skip 2 of this $.  parent().parent() is not a good combination either, but.. 

https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/master/assets/js/admin-scripts.js#L129-L135

You are requesting the DOM too many times with $. Cache it, and re-use. Use of $ would be down to 3 and not 5 as now.

In same file, it seems you are asking for the same selector many times, cache them in the top of the function or document. Reduces $.

https://github.com/easydigitaldownloads/Easy-Digital-Downloads/blob/master/assets/js/admin-scripts.js#L129-L135

Here cache $(this) or use event.target.

The list is much longer. I'm not after promoting anything here simply because what I'm developing is kind of non-public, but in the code I created. I could have used around 400 - 500 codes. Maybe less to archive the same things. And performance boost 50 - 60 % and no $.

Anyway. You find similar things in all the Javascript files. I'm now back to my own Javascript world!

**Comment by Kflash** [comment]
@chriscct7 I took a benchmark test on jQuery hide() / show() now, and there is some performance issues here. Very obvious if you look at the results (I compared with my own code). 

Anyway. The bottleneck for jQuery seems to be the show() method, and not the hide() method.

**------------------------show/hide------------------------------
jquery#hide() x 531,816 ops/sec ±1.34% (91 runs sampled)
jquery#show() x 19,910 ops/sec ±111.01% (91 runs sampled)
Ugma#hide() x 3,885,864 ops/sec ±0.26% (97 runs sampled)
Ugma#show() x 1,972,245 ops/sec ±94.32% (93 runs sampled)**

**Comment by dmethvin** [comment]
@Kflash those are synthetic benchmarks and don't necessarily reflect the performance of any code in this repo. Taking a look at the "slowest" number there with `.show()`, it still manages about 20,000 per second or 5 *microseconds* per operation. Showing or hiding a few elements every few seconds isn't going to cause any performance problem if those reflect your usage.

Ultimately you need to test *this* code like @chriscct7 said and determine where the performance bottlenecks are in this code. And you really only need to do that if you are perceiving that there are perf bottlenecks in the first place. There's no reason to go looking for trouble where there is none.

For background, the reason jQuery is discussing this topic is because high-profile sites are writing bad code that *does* have obvious and horrible bottlenecks and not fixing the performance problems that arise. Something like `.show()` isn't just a `display: block` because of all the corner cases like not-yet-attached elements that will be affected by stylesheets once inserted, or showing a row in a table where `block` isn't the correct display style. 

We're trying to fix the problem from the jQuery end because historically it has been very difficult to get sites to address performance problems. That's hard to believe until you go to https://en.wikipedia.org/wiki/Barack_Obama?veaction=edit and wonder how anyone could ever  ship that. I actually find IE's profiler output easier to interpret for this case than Chrome's:

![capture](https://cloud.githubusercontent.com/assets/157858/6190624/2ee84b4c-b371-11e4-8807-cc3efd4056a9.PNG)

During the load of that page, it's thrashing around in a cycle where it modifies the DOM and asks the browser (indirectly through jQuery) what parts of the DOM are visible. *Thousands* of times, look at the counts. Even an operation that is just slightly slow will be really slow if you do it thousands of times. Just the three calls to parse the DOM take half a second, that must be one heck of a lump of markup!

So if this ticket was filed against that page, yes someone should fix it and you have the evidence it needs to be fixed plus the profiler output to show you what needs fixing. For this case I don't think you have evidence of a problem, at least not yet.

**Comment by Kflash** [comment]
@dmethvin I first off need to note that I actually ended up in this conversation in the first place. I had a peasefull life with my own code. When that is said, and about this performance issue.  Maybe this thread is not a correct place if this is a jQuery issue? If it is. It can be so simple that there are a lot of bad developers that cause this. But in the end, it has to be a kind of a *chain reaction** inside jQuery that causes this to happen. One function or similiar that trigger the next chain, and then the other... etc. And you see the result.  Because a simple *display:block* can't cause this to happen. And there is no use of computed styles as I can see. So my guess is that this is not CSS related. Maybe worth to check things around the  iFrame in the display code?

I can investigate this a little, but in the first place it would be a good start to find other webpages with same bottlenecks and compare them all so try to isolate the problem. Any place we can continue this conversation?

**Comment by Kflash** [comment]
@dmethvin I first off need to note that I actually ended up in this conversation in the first place. I had a peasefull life with my own code. When that is said, and about this performance issue.  Maybe this thread is not a correct place if this is a jQuery issue? If it is. It can be so simple that there are a lot of bad developers that cause this. But in the end, it has to be a kind of a *chain reaction** inside jQuery that causes this to happen. One function or similiar that trigger the next chain, and then the other... etc. And you see the result.  Because a simple *display:block* can't cause this to happen. And there is no use of computed styles as I can see. So my guess is that this is not CSS related. Maybe worth to check things around the  iFrame in the display code?

I can investigate this a little, but in the first place it would be a good start to find other webpages with same bottlenecks and compare them all so try to isolate the problem. Any place we can continue this conversation?

**Comment by dmethvin** [comment]
@Kflash If this were my repo, I'd close the ticket until someone comes up with a real performance problem--specifically *not* a jsperf. :smiling_imp:

**Comment by pippinsplugins** [comment]
I do think we need to close this out and separate it (if needed) to specific issues. For example: `improve caching of selectors in edd-ajax.js`.

Generalized issues like this don't really go any where. Thank you everyone for your comments and feedback :)
